addons/binding/org.openhab.binding.plclogo/src/main/java/org/openhab/binding/plclogo/internal/PLCLogoClient.java

```diff
@@ -0,0 +1,201 @@
+     * @param RemoteTSAP Remote TSAP for the connection
+     */
+    @Override
+    public void SetConnectionParams(String Address, int LocalTSAP, int RemoteTSAP) {
+        address = Address;
+        super.SetConnectionParams(Address, LocalTSAP, RemoteTSAP);
+    }
+
+    /**
+     * Disconnects a client from a PLC
+     */
+    @Override
+    public synchronized void Disconnect() {
+        super.Disconnect();
+    }
+
+    /**
+     * Reads a data area from a PLC
+     *
+     * @param Area S7 Area ID. Can be S7AreaPE, S7AreaPA, S7AreaMK, S7AreaDB, S7AreaCT or S7AreaTM
+     * @param DBNumber S7 data block number
+     * @param Start First position within data block read from
+     * @param Amount Number of words to read
+     * @param WordLength Length of single word. Can be S7WLBit, S7WLByte, S7WLCounter or S7WLTimer
+     * @param Data Buffer to read into
+     * @return Zero on success, error code otherwise
+     */
+    @Override
+    public synchronized int ReadArea(int Area, int DBNumber, int Start, int Amount, int WordLength, byte[] Data) {
+        if (LastError != 0) {
+            logger.debug("Reconnect during read from {}: {}", address, ErrorText(LastError));
+            Disconnect();
+        }
+        if (!Connected) {
+            Connect();
+        }
+
+        final int packet = Math.min(Amount, 1024);
+        int offset = packet;
+
+        int retry = 0;
+        int result = -1;
+        do {
+            // read first portion directly to data
+            result = super.ReadArea(Area, DBNumber, Start, packet, WordLength, Data);
+            while ((result == 0) && (offset < Amount)) {
+                byte buffer[] = new byte[Math.min(Amount - offset, packet)];
+                result = super.ReadArea(Area, DBNumber, offset, buffer.length, WordLength, buffer);
+                System.arraycopy(buffer, 0, Data, offset, buffer.length);
+                offset = offset + buffer.length;
+            }
+
+            if (retry == MAX_RETRY_NUMBER) {
+                logger.warn("Giving up reading from {} after {} retries.", address, MAX_RETRY_NUMBER);
+                break;
+            }
+
+            if (result != 0) {
+                logger.warn("Reconnect during read from {}: {}", address, ErrorText(result));
+                retry = retry + 1;
+                Disconnect();
+                Connect();
+            }
+        } while (result != 0);
+
+        return result;
+    }
+
+    /**
+     * Reads a data block area from a PLC
+     *
+     * @param DBNumber S7 data block number
+     * @param Start First position within data block read from
+     * @param Amount Number of words to read
+     * @param WordLength Length of single word. Can be S7WLBit, S7WLByte, S7WLCounter or S7WLTimer
+     * @param Data Buffer to read into
+     * @return Zero on success, error code otherwise
+     */
+    public int ReadDBArea(int DBNumber, int Start, int Amount, int WordLength, byte[] Data) {
```

Comment:
methods should start with a lower level, please try to follow that here as well as far as Moka allows